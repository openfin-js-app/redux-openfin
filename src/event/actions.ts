import {EventConstant, appEvents, windowEvents} from './constants';

export type ActionDict = {
    [name:string]:EventConstant,
}

const appEventDictByName:ActionDict = {};
const appEventDictByType:ActionDict = {};
const windowEventDictByName:ActionDict = {};
const windowEventDictByType:ActionDict = {};

appEvents.forEach(oneEvent => {
    appEventDictByName[oneEvent.name] = oneEvent;
    appEventDictByType[oneEvent.type] = oneEvent;
});

windowEvents.forEach(oneEvent => {
    windowEventDictByName[oneEvent.name] = oneEvent;
    windowEventDictByType[oneEvent.type] = oneEvent;
});

export default {
    appEventDictByName,
    appEventDictByType,
    windowEventDictByName,
    windowEventDictByType,
};