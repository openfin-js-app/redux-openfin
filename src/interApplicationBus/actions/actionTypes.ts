import { makeReqType, makeResType } from '../../utils/makeType';

export const PUBLISH = makeReqType('PUBLISH');
export const PUBLISH_RES = makeResType('PUBLISH_RES');

export const SUBSCRIBE = makeReqType('SUBSCRIBE');
export const SUBSCRIBE_RES = makeResType('SUBSCRIBE_RES');
