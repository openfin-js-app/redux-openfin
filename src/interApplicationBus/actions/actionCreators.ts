import {Action} from 'redux-actions';
import createFSA, {ActionCreator} from '../../utils/createFSA';
import * as types from "../types";
import * as actionTypes from "./actionTypes";

export const publish:ActionCreator<types.PublishPayload> = createFSA<types.PublishPayload>(
        actionTypes.PUBLISH,
        (options:types.PublishPayload) => <any>{options}
);

export const subscribe:ActionCreator<types.SubscribePayload> = createFSA<types.SubscribePayload>(
        actionTypes.SUBSCRIBE,
        (options:types.SubscribePayload) => <any>{options}
);