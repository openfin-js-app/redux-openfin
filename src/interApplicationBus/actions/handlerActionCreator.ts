import {Action} from 'redux-actions';
import createFSA, {ActionCreator} from "../../utils/createFSA";
import * as types from "../types";
import * as actionTypes from "./actionTypes";

export const publishRes:ActionCreator<types.PublishResPayload> = createFSA<types.PublishResPayload>(
    actionTypes.PUBLISH_RES,
    (options:types.PublishResPayload) => <any>{options}
);

export const subscribeRes:ActionCreator<types.SubscribeResPayload> = createFSA<types.SubscribeResPayload>(
    actionTypes.SUBSCRIBE_RES,
    (options:types.SubscribeResPayload) => <any>{options}
);