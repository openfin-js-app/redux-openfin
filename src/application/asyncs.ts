import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import createAsyncFun from '../utils/createAsyncFun'

import {
    NEW_APPLICAITON_ERROR_MSG,
    RESTART_ERROR_MSG,
    CLOSE_ERROR_MSG,
} from './types';

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#Application
export async function newApplication(action:Action<types.NewApplicationPayload>):Promise<Action<types.NewApplicationResPayload>>{
    const  options  = action.payload;
    return createAsyncFun<types.NewApplicationPayload,types.NewApplicationResPayload>(
        action,
        NEW_APPLICAITON_ERROR_MSG,
        handlerActions.newApplicationRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let app = new fin.desktop.Application(options,
                (app:any)=>{
                    const responseAction = resActionCreator({app});
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#.getCurrent

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getWindow

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#.wrap

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#addEventListener

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#close
export async function close(action:Action<types.ClosePayload>):Promise<Action<types.CloseResPayload>>{
    const { force } = action.payload;
    return createAsyncFun<types.ClosePayload,types.CloseResPayload>(
        action,
        CLOSE_ERROR_MSG,
        handlerActions.closeRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let application = fin.desktop.Application.getCurrent();
            application.close(force,
                ()=>{
                    const responseAction = resActionCreator({});
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getChildWindows

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getInfo

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getShortcuts

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getTrayIconInfo

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getZoomLevel

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#isRunning

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#removeEventListener

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#restart
export async function restart(action:Action<types.RestartPayload>):Promise<Action<types.RestartResPayload>>{
    return createAsyncFun<types.RestartPayload,types.NewApplicationResPayload>(
        action,
        RESTART_ERROR_MSG,
        handlerActions.restartRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let application = fin.desktop.Application.getCurrent();
            application.restart(
                ()=>{
                    const responseAction = resActionCreator({});
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#run

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#scheduleRestart

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#setShortcuts

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#setTrayIcon

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#setZoomLevel

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#terminate
