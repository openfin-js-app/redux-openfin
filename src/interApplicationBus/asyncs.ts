import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import wrapAsyncFun from '../utils/wrapAsyncFun'

export async function publish(action:Action<types.PublishPayload>):Promise<Action<types.PublishResPayload>>{
    const  { topic, message }  = action.payload;
    return wrapAsyncFun<types.PublishPayload,types.PublishResPayload>(
        action,
        handlerActions.publishRes,
        async (fin)=>{
            await fin.InterApplicationBus.publish(topic, message);
            return handlerActions.publishRes({});
        }
    );
}

export async function subscribe(action:Action<types.SubscribePayload>):Promise<Action<types.SubscribeResPayload>>{
    const  { source, topic, listener }  = action.payload;
    return wrapAsyncFun<types.SubscribePayload,types.SubscribeResPayload>(
        action,
        handlerActions.subscribeRes,
        async (fin)=>{
            await fin.InterApplicationBus.subscribe(source, topic, listener);
            return handlerActions.subscribeRes({});
        }
    );
}
