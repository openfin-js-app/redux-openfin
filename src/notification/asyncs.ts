import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import {initState} from '../init';

import createAsyncFun from '../utils/createAsyncFun'

import uuid from '../utils/uuid';

import {
    CREATE_NOTIFICATION_ERROR_MSG,
    NOTIFICATION_GET_CURRENT_ERROR_MSG,
    NOTIFICATION_CLOSE_ERROR_MSG,
    NOTIFICATION_SEND_MSG_ERROR_MSG,
    NOTIFICATION_SEND_MSG_TO_APP_ERROR_MSG,
} from './types';

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#Notification
export async function createNotification(action:Action<types.CreateNotificationPayload>):Promise<Action<types.CreateNotificationResPayload>>{
    const  options  = action.payload;

    let userId = action.payload.userId;
    if (!userId){
        userId = uuid();
    }

    let finOptions:any = {...options};
    delete finOptions.userId;
    if (initState.store){
        const dispatch = initState.store.dispatch;
        finOptions.onClick = ()=>{
            dispatch(handlerActions.notificationOnClickRes({userId}));
        };
        finOptions.onClose = ()=>{
            dispatch(handlerActions.notificationOnCloseRes({userId}));
        };
        finOptions.onDismiss = ()=>{
            dispatch(handlerActions.notificationOnDismissRes({userId}));
        };
        finOptions.onError = (reason,errorObj)=>{
            dispatch(handlerActions.notificationOnErrorRes({userId,reason,errorObj}));
        };
        finOptions.onMessage = (message)=>{
            dispatch(handlerActions.notificationOnMsgRes({userId,message}));
        };
        finOptions.onShow = (successObj)=>{
            dispatch(handlerActions.notificationOnShowRes({userId,successObj}));
        };
    }

    return createAsyncFun<types.CreateNotificationPayload,types.CreateNotificationResPayload>(
        action,
        CREATE_NOTIFICATION_ERROR_MSG,
        handlerActions.createNotificationRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let notification = new fin.desktop.Notification(finOptions,
                ()=>{
                    const responseAction = resActionCreator({
                        userId,
                        notification
                    });
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#.getCurrent
export async function getCurrent(action:Action<types.NotificationGetCurrentPayload>):Promise<Action<types.NotificationGetCurrentResPayload>>{
    return createAsyncFun<types.NotificationGetCurrentPayload,types.NotificationGetCurrentResPayload>(
        action,
        NOTIFICATION_GET_CURRENT_ERROR_MSG,
        handlerActions.getCurrentRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({notification})
            succCb(responseAction);
        }
    )
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#close
export async function close(action:Action<types.NotificationClosePayload>):Promise<Action<types.NotificationCloseResPayload>>{
    return createAsyncFun<types.NotificationClosePayload,types.NotificationCloseResPayload>(
        action,
        NOTIFICATION_CLOSE_ERROR_MSG,
        handlerActions.closeRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({})
            notification.close(()=>{
                succCb(responseAction);
            });
        }
    )
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#sendMessage
export async function sendMessage(action:Action<types.NotificationSendMsgPayload>):Promise<Action<types.NotificationSendMsgResPayload>>{

    const { payload: {
        message
    }} = action;

    return createAsyncFun<types.NotificationSendMsgPayload,types.NotificationSendMsgResPayload>(
        action,
        NOTIFICATION_SEND_MSG_ERROR_MSG,
        handlerActions.sendMessageRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({})
            notification.sendMessage(message,()=>{
                succCb(responseAction);
            });
        }
    )
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#sendMessageToApplication
export async function sendMessageToApplication(action:Action<types.NotificationSendMsgToAppPayload>):Promise<Action<types.NotificationSendMsgToAppResPayload>>{

    const { payload: {
        message
    }} = action;

    return createAsyncFun<types.NotificationSendMsgToAppPayload,types.NotificationSendMsgToAppResPayload>(
        action,
        NOTIFICATION_SEND_MSG_TO_APP_ERROR_MSG,
        handlerActions.sendMessageToApplicationRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({})
            notification.sendMessageToApplication(message,()=>{
                succCb(responseAction);
            });
        }
    )
}