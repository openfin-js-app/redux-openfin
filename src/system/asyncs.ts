import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import createAsyncFun from '../utils/createAsyncFun'

import {
    GET_MACHINE_ID_ERROR_MSG,
    GET_DEVICE_USER_ID_ERROR_MSG,
    GET_MONITOR_INFO_ERROR_MSG,
    GET_VERSION_ERROR_MSG,
    GET_HOST_SPECS_ERROR_MSG,
    GET_ENVIRONMENT_VARIABLE_ERROR_MSG,
    CLEAR_CACHE_ERROR_MSG,
} from './types';

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.System.html#.getDeviceId
export async function getDeviceId(action:Action<types.GetMachineIdPayload>):Promise<Action<types.GetMachineIdResPayload>>{
    return createAsyncFun<types.GetMachineIdPayload,types.GetMachineIdResPayload>(
        action,
        GET_MACHINE_ID_ERROR_MSG,
        handlerActions.getMachineIdRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            fin.desktop.System.getMachineId(
                (id:string)=>{
                    const responseAction = resActionCreator({id});
                    succCb(responseAction);
                },
                errCb
            );
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.System.html#.getDeviceUserId
export async function getDeviceUserId(action:Action<types.GetDeviceUserIdPayload>):Promise<Action<types.GetDeviceUserIdResPayload>>{
    return createAsyncFun<types.GetDeviceUserIdPayload,types.GetDeviceUserIdResPayload>(
        action,
        GET_DEVICE_USER_ID_ERROR_MSG,
        handlerActions.getDeviceUserIdRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            fin.desktop.System.getDeviceUserId(
                (id:string)=>{
                    const responseAction = resActionCreator({id});
                    succCb(responseAction);
                },
                errCb
            );
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.System.html#.getMonitorInfo
export async function getMonitorInfo(action:Action<types.GetMonitorInfoPayload>):Promise<Action<types.GetMonitorInfoResPayload>>{
    return createAsyncFun<types.GetMonitorInfoPayload,types.GetMonitorInfoResPayload>(
        action,
        GET_MONITOR_INFO_ERROR_MSG,
        handlerActions.getMonitorInfoRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            fin.desktop.System.getMonitorInfo(
                (monitorInfo:any)=>{
                    const responseAction = resActionCreator(monitorInfo);
                    succCb(responseAction);
                },
                errCb
            );
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.System.html#.getVersion
export async function getVersion(action:Action<types.GetVersionPayload>):Promise<Action<types.GetVersionResPayload>>{
    return createAsyncFun<types.GetVersionPayload,types.GetVersionResPayload>(
        action,
        GET_VERSION_ERROR_MSG,
        handlerActions.getVersionRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            fin.desktop.System.getVersion(
                (version:string)=>{
                    const responseAction = resActionCreator({version});
                    succCb(responseAction);
                },
                errCb
            );
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.System.html#.getHostSpecs
export async function getHostSpecs(action:Action<types.GetHostSpecsPayload>):Promise<Action<types.GetHostSpecsResPayload>>{
    return createAsyncFun<types.GetHostSpecsPayload,types.GetHostSpecsResPayload>(
        action,
        GET_HOST_SPECS_ERROR_MSG,
        handlerActions.getHostSpecsRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            fin.desktop.System.getHostSpecs(
                (info:types.GetHostSpecsResPayload)=>{
                    const responseAction = resActionCreator(info);
                    succCb(responseAction);
                },
                errCb
            );
        }
    );
}


//http://cdn.openfin.co/jsdocs/beta/fin.desktop.System.html#.getEnvironmentVariable
export async function getEnvironmentVariable(action:Action<types.GetEnvironmentVariablePayload>)
    :Promise<Action<types.GetEnvironementVariableResPaylod>>
{
    const { env } = action.payload;
    return createAsyncFun<types.GetEnvironmentVariablePayload,types.GetEnvironementVariableResPaylod>(
        action,
        GET_ENVIRONMENT_VARIABLE_ERROR_MSG,
        handlerActions.getEnvironmentVariableRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            fin.desktop.System.getEnvironmentVariable(
                env,
                (value:string)=>{
                    const responseAction = resActionCreator({env,value});
                    succCb(responseAction);
                },
                errCb
            );
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.System.html#.clearCache
export async function clearCache(action:Action<types.ClearCachePayload>):Promise<Action<types.ClearCacheResPayload>>{
    const  options  = action.payload;
    return createAsyncFun<types.ClearCachePayload,types.ClearCacheResPayload>(
        action,
        CLEAR_CACHE_ERROR_MSG,
        handlerActions.clearCacheRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            fin.desktop.System.clearCache(
                options,
                ()=>{
                    const responseAction = resActionCreator({});
                    succCb(responseAction);
                },
                errCb
            );
        }
    );

}