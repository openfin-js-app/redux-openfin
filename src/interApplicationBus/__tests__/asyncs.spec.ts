jest.mock('../../utils/createAsyncFun');
const wrapAsyncFun = require('../../utils/wrapAsyncFun');

import * as actions from '../actions';
import * as asyncs  from '../asyncs';

const getDefaultPromise = ()=>{
    return new Promise(((resolve, reject) => {
        resolve({});
    }))
}

describe('InterApplication asyncs',()=>{

    afterEach(()=>{
        jest.resetAllMocks();
    });

    it('publish',async ()=>{

        const fin = {
            InterApplicationBus : {
                publish:jest.fn(getDefaultPromise),
            }
        };

        wrapAsyncFun.default=jest.fn(
            async (action,resActionCreator,finCb)=>{
                return await finCb(fin,action,resActionCreator);
            }
        );

        await asyncs.publish(actions.publish({
            topic:'topic',
            message:'message',
        }));

        expect(fin.InterApplicationBus.publish).toHaveBeenCalled();
        expect(wrapAsyncFun.default).toHaveBeenCalled();

    })

    it('subscribe',async ()=>{

        const fin = {
            InterApplicationBus : {
                subscribe:jest.fn(getDefaultPromise),
            }
        };

        wrapAsyncFun.default=jest.fn(
            async (action,resActionCreator,finCb)=>{
                return await finCb(fin,action,resActionCreator);
            }
        );

        await asyncs.subscribe(actions.subscribe({
            senderUuid:'senderUuid',
            topic:'topic',
            listener:(message,uuid,name)=>{},
        } as any));

        expect(fin.InterApplicationBus.subscribe).toHaveBeenCalled();
        expect(wrapAsyncFun.default).toHaveBeenCalled();

    })


});