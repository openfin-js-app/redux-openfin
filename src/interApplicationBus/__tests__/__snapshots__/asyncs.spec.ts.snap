// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterApplication asyncs publish 1`] = `
Object {
  "__esModule": true,
  "default": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "error": undefined,
          "payload": Object {
            "options": Object {
              "message": "message",
              "topic": "topic",
            },
          },
          "type": "REDUX_OPENFIN_REQ::PUBLISH",
        },
        "OpenFin API call InterApplicationBus.publish failed.",
        [Function],
        [Function],
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
}
`;

exports[`InterApplication asyncs subscribe 1`] = `
Object {
  "__esModule": true,
  "default": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "error": undefined,
          "payload": Object {
            "options": Object {
              "listener": [Function],
              "senderUuid": "senderUuid",
              "topic": "topic",
            },
          },
          "type": "REDUX_OPENFIN_REQ::SUBSCRIBE",
        },
        "OpenFin API call InterApplicationBus.subscribe failed.",
        [Function],
        [Function],
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
}
`;
